src/v/archival/ntp_archiver_service.cc

```diff
@@ -384,8 +384,10 @@ ss::future<ntp_archiver::batch_result> ntp_archiver::wait_all_scheduled_uploads(
         if (_partition->archival_meta_stm()) {
             retry_chain_node rc_node(
               _manifest_upload_timeout, _initial_backoff, &parent);
-            auto error = co_await _partition->archival_meta_stm()->add_segments(
+            auto fut = _partition->archival_meta_stm()->add_segments(
               _manifest, rc_node);
+            auto deadline = ss::lowres_clock::now() + _manifest_upload_timeout;
+            auto error = co_await ss::with_timeout(deadline, std::move(fut));
```

Comment:
I think it would be more uniform to put `with_timeout` inside the`add_segments` function and return an error code in case of timeout.

Reply:
+1, maybe push the with_timeout call all the way down to where raft->replicate is called?  That way it's clear which part of the process is subject to the timeout, and if anyone else ever calls add_segments then they are safe too.